service_report_generator.py

```diff
@@ -0,0 +1,256 @@
+
+            print("\n---\n\n", file=self.fout)
+
+    def _print_test_header(self, test_name, description, rationale):
+        print("### %s\n\n" % test_name, file=self.fout)
+        if description:
+            print("**Description:** %s\n" % description, file=self.fout)
+        if rationale:
+            print("**Rationale:** %s\n" % rationale, file=self.fout)
+
+    def _print_test_result_tables(self, test_name):
+        for status in STATUSES_TO_LIST:
+            if self._test_results_include_status(test_name, status):
+                print("#### %s_%s\n\n" % (test_name, status), file=self.fout)
+                print("Resource Name | Metadata", file=self.fout)
+                print("------------ | -------------", file=self.fout)
+
+                for result in self.test_results[test_name]:
+                    if result["status"] == status:
+                        print("%s | %s" % (
+                            self._extract_resource_name(result['name']),
+                            self._format_metadata(result['metadata'])
+                        ), file=self.fout)
+
+                print("\n\n", file=self.fout)
+
+    def _print_status_table(self):
+        print("Status Name | Meaning", file=self.fout)
+        print("------------ | -------------", file=self.fout)
+        print("fail | Critical test failure that should never happen e.g. publicly "
+              "accessible DB snapshot, user without MFA in prod.", file=self.fout)
+        print("warn | Non-critical test result like an unexpected test failure "
+              "(xpass, xfail, skip). Examples: S3 bucket not tagged as public, "
+              "error fetching resource from the AWS API, test skipped due to it not applying", file=self.fout)
+        print("err | Error fetching an resource from AWS.", file=self.fout)
+        print("\n\n", file=self.fout)
+
+    def _extract_resource_name(self, name):
+        # "test_something[resource-name]" -> "resource-name"
+        return name.split("[")[-1][0:-1]
+
+    def _format_metadata(self, metadata):
+        """
+        Formats the metadata dictionary to a string that is somewhat readable in a markdown table.
+
+        >>> _format_metadata({'foo': 'bar'})
+        "foo: bar"
+        >>> _format_metadata({'VpcId': '1234', 'GroupName': 'ssh-only'})
+        "VpcId: 1234 - GroupName: ssh-only"
+        """
+        return ''.join(["{}: {} - ".format(k, v) for k, v in metadata.items()])[0:-3]
+
+    # test results include at least one with status of
+    def _test_results_include_status(self, test_name, status):
+        return bool(self.test_status_counter[test_name+'_'+status])
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(description=__doc__)
+
+    parser.add_argument('-jo', '--json-out', default='service-report.json',
+                        help='Service json output filename.')
+
+    parser.add_argument('-mo', '--markdown-out', default='service-report.md',
```

Comment:
nit: I think `-jo` and `-mo` usually imply `-j -o` and  `-m -o` respectively for flags, so aliases with two dashes `--jo` and `--mo` would keep with unix convention

https://serverfault.com/questions/387935/whats-the-difference-betwen-the-single-dash-and-double-dash-flags-on-shell-comm

Reply:
👍 fixed